// werift/lib/dtls/src/handshake/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HandshakeType = void 0;
var HandshakeType;
(function (HandshakeType) {
    HandshakeType[HandshakeType["hello_request_0"] = 0] = "hello_request_0";
    HandshakeType[HandshakeType["client_hello_1"] = 1] = "client_hello_1";
    HandshakeType[HandshakeType["server_hello_2"] = 2] = "server_hello_2";
    HandshakeType[HandshakeType["hello_verify_request_3"] = 3] = "hello_verify_request_3";
    HandshakeType[HandshakeType["certificate_11"] = 11] = "certificate_11";
    HandshakeType[HandshakeType["server_key_exchange_12"] = 12] = "server_key_exchange_12";
    HandshakeType[HandshakeType["certificate_request_13"] = 13] = "certificate_request_13";
    HandshakeType[HandshakeType["server_hello_done_14"] = 14] = "server_hello_done_14";
    HandshakeType[HandshakeType["certificate_verify_15"] = 15] = "certificate_verify_15";
    HandshakeType[HandshakeType["client_key_exchange_16"] = 16] = "client_key_exchange_16";
    HandshakeType[HandshakeType["finished_20"] = 20] = "finished_20";
})(HandshakeType = exports.HandshakeType || (exports.HandshakeType = {}));
