// werift/lib/sctp/src/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SCTP_STATE = exports.WEBRTC_PPID = void 0;
var WEBRTC_PPID;
(function (WEBRTC_PPID) {
    WEBRTC_PPID[WEBRTC_PPID["DCEP"] = 50] = "DCEP";
    WEBRTC_PPID[WEBRTC_PPID["STRING"] = 51] = "STRING";
    WEBRTC_PPID[WEBRTC_PPID["BINARY"] = 53] = "BINARY";
    WEBRTC_PPID[WEBRTC_PPID["STRING_EMPTY"] = 56] = "STRING_EMPTY";
    WEBRTC_PPID[WEBRTC_PPID["BINARY_EMPTY"] = 57] = "BINARY_EMPTY";
})(WEBRTC_PPID = exports.WEBRTC_PPID || (exports.WEBRTC_PPID = {}));
var SCTP_STATE;
(function (SCTP_STATE) {
    SCTP_STATE[SCTP_STATE["CLOSED"] = 1] = "CLOSED";
    SCTP_STATE[SCTP_STATE["COOKIE_WAIT"] = 2] = "COOKIE_WAIT";
    SCTP_STATE[SCTP_STATE["COOKIE_ECHOED"] = 3] = "COOKIE_ECHOED";
    SCTP_STATE[SCTP_STATE["ESTABLISHED"] = 4] = "ESTABLISHED";
    SCTP_STATE[SCTP_STATE["SHUTDOWN_PENDING"] = 5] = "SHUTDOWN_PENDING";
    SCTP_STATE[SCTP_STATE["SHUTDOWN_SENT"] = 6] = "SHUTDOWN_SENT";
    SCTP_STATE[SCTP_STATE["SHUTDOWN_RECEIVED"] = 7] = "SHUTDOWN_RECEIVED";
    SCTP_STATE[SCTP_STATE["SHUTDOWN_ACK_SENT"] = 8] = "SHUTDOWN_ACK_SENT";
})(SCTP_STATE = exports.SCTP_STATE || (exports.SCTP_STATE = {}));
