// werift/lib/webrtc/src/media/track.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaStream = exports.MediaStreamTrack = void 0;
const rx_mini_1 = __importDefault(require("rx.mini"));
const uuid_1 = require("uuid");
const src_1 = require("../../../rtp/src");
const helper_1 = require("../helper");
class MediaStreamTrack extends helper_1.EventTarget {
    constructor(props) {
        super();
        this.uuid = uuid_1.v4();
        this.remote = false;
        /**todo impl */
        this.enabled = true;
        this.onReceiveRtp = new rx_mini_1.default();
        this.stopped = false;
        this.muted = true;
        this.stop = () => {
            this.stopped = true;
            this.muted = true;
            this.onReceiveRtp.complete();
        };
        this.writeRtp = (rtp) => {
            if (this.remote)
                throw new Error("this is remoteTrack");
            if (!this.codec || this.stopped)
                return;
            const packet = Buffer.isBuffer(rtp) ? src_1.RtpPacket.deSerialize(rtp) : rtp;
            packet.header.payloadType = this.codec.payloadType;
            this.onReceiveRtp.execute(packet);
        };
        Object.assign(this, props);
        this.onReceiveRtp.subscribe((rtp) => {
            this.muted = false;
            this.header = rtp.header;
        });
        this.label = `${this.remote ? "remote" : "local"} ${this.kind}`;
    }
}
exports.MediaStreamTrack = MediaStreamTrack;
class MediaStream {
    constructor(props) {
        this.tracks = [];
        Object.assign(this, props);
    }
    addTrack(track) {
        this.tracks.push(track);
    }
    getTracks() {
        return this.tracks;
    }
}
exports.MediaStream = MediaStream;
